packages/marshal/src/marshal.js

```diff
@@ -875,52 +899,63 @@ export function makeMarshal(
-    // not frozen during the recursion. Rather, the whole graph is
-    // hardened before being returned. Error objects are not
-    // potentially recursive, and so may be harmlessly hardened when
-    // they are produced.
-    //
-    // fullRevive can produce properties whose value is undefined,
-    // which a JSON.parse on a reviver cannot do. If a reviver returns
-    // undefined to JSON.parse, JSON.parse will delete the property
-    // instead.
-    //
-    // fullRevive creates and returns a new graph, rather than
-    // modifying the original tree in place.
-    //
-    // fullRevive may rely on rawTree being the result of a plain call
-    // to JSON.parse. However, it *cannot* rely on it having been
-    // produced by JSON.stringify on the replacer above, i.e., it
-    // cannot rely on it being a valid marshalled
-    // representation. Rather, fullRevive must validate that.
-    return function fullRevive(rawTree) {
+    /**
+     * We stay close to the algorithm at
+     * https://tc39.github.io/ecma262/#sec-json.parse , where
+     * fullRevive(harden(JSON.parse(str))) is like JSON.parse(str, revive))
+     * for a similar reviver. But with the following differences:
+     *
+     * Rather than pass a reviver to JSON.parse, we first call a plain
+     * (one argument) JSON.parse to get rawTree, and then post-process
+     * the rawTree with fullRevive. The kind of revive function
+     * handled by JSON.parse only does one step in post-order, with
+     * JSON.parse doing the recursion. By contrast, fullParse does its
+     * own recursion, enabling it to interpret ibids in the same
+     * pre-order in which the replacer visited them, and enabling it
+     * to break cycles.
+     *
+     * In order to break cycles, the potentially cyclic objects are
+     * not frozen during the recursion. Rather, the whole graph is
+     * hardened before being returned. Error objects are not
+     * potentially recursive, and so may be harmlessly hardened when
+     * they are produced.
+     *
+     * fullRevive can produce properties whose value is undefined,
+     * which a JSON.parse on a reviver cannot do. If a reviver returns
+     * undefined to JSON.parse, JSON.parse will delete the property
+     * instead.
+     *
+     * fullRevive creates and returns a new graph, rather than
+     * modifying the original tree in place.
+     *
+     * fullRevive may rely on rawTree being the result of a plain call
+     * to JSON.parse. However, it *cannot* rely on it having been
+     * produced by JSON.stringify on the replacer above, i.e., it
+     * cannot rely on it being a valid marshalled
+     * representation. Rather, fullRevive must validate that.
+     *
+     * @param {Encoding} rawTree must be hardened
+     */
+    function fullRevive(rawTree) {
       if (Object(rawTree) !== rawTree) {
         // primitives pass through
         return rawTree;
       }
+      // Assertions of the above to narrow the type.
```

Comment:
Should we replace all 
```js
if (Object(x) === x) {
```
with a call to an isPrimitive so we can declare typing implications for the true and false branches? Seems more structured than adding as `assert` to branches after the test.

Reply:
That's a good idea.  I wonder where such a function should live?

Reply:
I don't have a good answer either. It is strange that we don't have a natural place to put such things. Please go ahead an just put it in marshal.js for now with a TODO to move it to an appropriate place for reuse.

Reply:
Unfortunately, I couldn't find a way to type `isPrimitive` so that it actually worked as well as the explicit asserts.  Maybe someday we can write such a function.